File: server/models/product.js
const mongoose = require("mongoose");
const ratingSchema = require("./rating");

const productSchema = new mongoose.Schema({
  name: {
    type: String,
    required: true,
    trim: true,
  },

  description: {
    type: String,
    required: true,
    trim: true,
  },

  images: [
    {
      type: String,
      required: true,
      trim: true,
    },
  ],
  quantity: {
    type: Number,
    required: true,
  },

  price: {
    type: Number,
    required: true,
  },

  category: {
    type: String,
    required: true,
  },

  ratings: [ratingSchema],
});

const Product = mongoose.model("product", productSchema);

module.exports = { Product, productSchema }; // Export the model and schema for use in other files

File: lib/models/product.dart
import 'dart:convert';

import 'package:amcart/models/rating.dart';

class Product {
  final String name;
  final String description;
  final double quantity;
  final List<String> images;
  final String category;
  final double price;
  final String? id;
  final List<Rating>? rating;
  Product({
    required this.name,
    required this.description,
    required this.quantity,
    required this.images,
    required this.category,
    required this.price,
    this.id,
    this.rating,
  });

  Map<String, dynamic> toMap() {
    return {
      'name': name,
      'description': description,
      'quantity': quantity,
      'images': images,
      'category': category,
      'price': price,
      'id': id,
      'ratings': rating,
    };
  }

  factory Product.fromMap(Map<String, dynamic> map) {
    return Product(
      name: map['name'] ?? '',
      description: map['description'] ?? '',
      quantity: map['quantity']?.toDouble() ?? 0.0,
      images: List<String>.from(map['images']),
      category: map['category'] ?? '',
      price: map['price']?.toDouble() ?? 0.0,
      id: map['_id'],
      // If the map contains a 'ratings' key, and the value is not null,
      // then create a List<Rating> from the value, which is a List of Maps.
      // For each Map in the List, create a Rating object from it using the
      // Rating.fromMap factory, and add it to the List.
      // If the 'ratings' key does not exist in the map, or if the value is null,
      // then set the ratings parameter to null.
      rating: map['ratings'] != null
          ? List<Rating>.from(
              map['ratings']?.map<Rating>((x) => Rating.fromMap(x)))
          : null,
    );
  }

  String toJson() => json.encode(toMap());

  factory Product.fromJson(String source) =>
      Product.fromMap(json.decode(source));
}
